api/Makefile

```diff
@@ -11,22 +11,11 @@ clean:
 	@test ! -e bin || rm -r bin
 	@test ! -e vendor || rm -rf vendor
 
-.PHONY: setup
-setup:
-	@echo "Setting up tools..."
-	@test -x $(shell go env GOPATH)/bin/golangci-lint || \
-		curl -sSfL https://raw.githubusercontent.com/golangci/golangci-lint/v1.46.2/install.sh | sh -s -- -b $(shell go env GOPATH)/bin v1.46.2
-
 .PHONY: fmt
 fmt:
 	@echo "Formatting code..."
 	gofmt -s -w ${SRC_ROOT}
 
-.PHONY: lint
-lint: setup
-	@echo "Linting code..."
-	golangci-lint -v run --timeout 3m $(if $(filter true,$(fix)),--fix,)
```

Comment:
Is it required to remove it? What if I want to lint the code without committing the changes? Would it be better to just integrate `make lint` as a command in the pre-commit hook?

Reply:
I was asking myself this. I rarely run `make lint` locally and usually the CI is the first to catch the problem (which is now expected to be sped up with the pre-commit hook). Anyway, I added it back.

But I went in the other direction, to let the pre-commit hook be run in the `--all-files` mode (as opposed running only if something changed, which is the behaviour of the automatic hook) in the make target. This is so that the set up is nicely contained in `.pre-commit-config.yaml` rather than installing `golangci-lint` externally.

```Make
.PHONY: lint
lint:
	@pre-commit run golangci-lint-experiment --all-files
```

Reply:
> But I went in the other direction

I'm not a fan of this, would like to hear the opinions of other folks. I'm used to seeing `Makefile` as a mission control to  "rule them all" of the repo – i.e. to contain all the recipes of the repository. In my opinion, it's the make file that should contain the configuration of the recipe, where the pre-commit hook only helps to run this command and the prescribed time (pre-commit). In other words, make commands should work perfectly fine without the pre-commit, where the opposite is not guaranteed.

Reply:
I agree it's unconventional to have a "pre-commit" tool drive the Makefile scripts. But I think it's more than that (the name is less helpful - in this case we are talking about a `manual` hook): https://pre-commit.com/#advanced

Especially for something like linting, I doubt that we would invoke this manually as often. Secondly, for some of our components, the CI doesn't rely on the same mechanism either (we use the relevant GitHub action as opposed to `make lint`). So my preference is the following:
1. Remove the explicit lint target and simply rely on the hooks
2. Accept `pre-commit` as our standard dev tooling and be ok with it being invoked from the Makefile
3. Convert most/all pre-commit hooks to local and let the Makefile rule them all (what you're suggesting). I am less enthused by this because we have to handwrite the script for installing the linter, testing if its version matches, etc. than leveraging something that's declarative. But no strong opinion on this.

Other @gojek/turing-dev, any thoughts?

Reply:
I use `make lint` sometimes so it'll be less convenient to see it go and I see how it's weird to be seeing pre-commit hooks in a Makefile. However it's also true that this means we'll have to 'handwrite the script for installing the linter, testing if its version matches'. Not too sure what the usual convention to address this so I don't really have a hard preference.

Reply:
Thanks for the input, @deadlycoconuts !

> Would there an issue keeping the make lint targets without pre-commit hooks and then accepting that we'll use the pre-commit hooks when committing as part our standard dev workflow? Only issue I can see with this is that we'll have 2 locations to maintain linters/versions, etc.

Yes, I considered this too but was not too keen on it because of the downside you've highlighted. Let's call it. # 4 :) 

@romanwozniak it seems neither Ziyi nor I have a strong opinion on this. And also this is something that may be applicable to all our repos. So, let's go with your call - are you still in favor of option # 3 ? And if so, should we be consistent about all the hooks being local, linking to the respective Make targets ?

Reply:
Hello @romanwozniak, any thoughts after the recent comments on this thread?

Reply:
I still prefer option 3 the most, but it's on you to decide what works the best.

>  And if so, should we be consistent about all the hooks being local, linking to the respective Make targets ?

Generally, I like that, except that for the front-end module, npm/yarn effectively replaces Makefile with `package.json > scripts` (meaning, that I don't think to have a Makefile for the UI module, because it would just duplicate yarn script)

Reply:
Cool. I've changed all except the helm-docs generation (which is unlikely to be run locally using `make`) as local hooks.